server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -634,73 +635,82 @@ public String toString() {
-                        } else { // corruption has happened on the way to replica
-                            RemoteTransportException exception = new RemoteTransportException("File corruption occurred on recovery but " +
-                                    "checksums are ok", null);
-                            exception.addSuppressed(e);
-                            logger.warn(() -> new ParameterizedMessage(
-                                    "{} Remote file corruption on node {}, recovering {}. local checksum OK",
-                                    shardId, request.targetNode(), md), corruptIndexException);
-                            throw exception;
+                    long position = 0;
+                    int bytesRead;
+                    final InputStream in = new InputStreamIndexInput(indexInput, md.length());
+                    while ((bytesRead = in.read(buffer, 0, buffer.length)) != -1) {
+                        final BytesArray content = new BytesArray(buffer, 0, bytesRead);
+                        final boolean lastChunk = position + content.length() == md.length();
+                        final long seqId = ++seqIdGenerator;
+                        // allow up to 2*maxConcurrentFileChunks buffering requests on the target
+                        requestSeqIdTracker.waitForOpsToComplete(seqId - (2 * maxConcurrentFileChunks));
+                        sendingTickets.acquire();
+                        boolean submitted = false;
+                        try {
+                            if (error.get() != null) {
+                                return; // abort and error will be handled in the finally block
+                            }
+                            cancellableThreads.checkForCancel();
+                            recoveryTarget.writeFileChunk(md, position, content, lastChunk, translogOps.get(), e -> {
+                                if (e != null) {
+                                    error.compareAndSet(null, Tuple.tuple(md, e));
+                                }
+                                requestSeqIdTracker.markSeqNoAsCompleted(seqId);
+                                sendingTickets.release();
+                            });
+                            position += content.length();
+                            submitted = true;
+                        } finally {
+                            if (submitted == false) {
+                                requestSeqIdTracker.markSeqNoAsCompleted(seqId);
+                                sendingTickets.release();
+                            }
                         }
-                    } else {
-                        throw e;
                     }
+                } catch (IOException e) {
+                    handleErrorOnSendFiles(store, md, e);
                 }
             }
         } finally {
-            store.decRef();
+            try {
+                sendingTickets.acquire(maxConcurrentFileChunks);
```

Comment:
I wonder if there are issues here with waiting for everything to complete when there is cancellation / exceptions going on. We don't particularly care about pending outgoing requests when we terminate exceptionally, as when those return we won't do anything with them anyway.

I think we should execute this here only on the happy path